Complete rework of the UCT BTL

hjelmn opened the pull request:
This PR contains a number of changes related to the UCT btl, these include:

 - Allow alternate connection methods for wiring up connect-to-endpoint transports. This includes using tcp to wire up ib. This may be needed depending on the level of UD support provided by a transport.
 - Clean up of modex code to improve readability.
 - Fix for a connection management issue that could cause accessing already-freed resources. This is separate from the other CLs but was discovered in the middle of the rework. I can take the time to extract if if needed.
 - Split the initialization code into a discovery and initialization phase. This is used to 1) check what memory domains are available, and 2) figure out network defaults. This code enables support for new MCA variables that allow controlling memory-domain specific behavior (since each memory domain may be a different btl).

hjelmn commented:
I know this will be a bit difficult to review. Looking at how I might be able to make it more reviewable. The first and second commit might be able to go in on their own but I improved the "used a different tl" logic in another commit. Was kind of hard working on this with the VMs being deleted once every 5 days and doing more defensive commits than having a good breakdown. I can probably split off creating the new logic vs removing the old. Will see.

What I really need is some verification of this on Mellanox hardware. This code greatly improves what we need for our RoCE but I don't want to break IB and I don't have anything to test with.

arun-chandran-edarath commented:
> What I really need is some verification of this on Mellanox hardware. This code greatly improves what we need for our RoCE but I don't want to break IB and I don't have anything to test with.

@hjelmn I am interested in collaborating and can test this on our IB clusters. How would you like to proceed? You can reach me at [arun.chandran@amd.com](mailto:arun.chandran@amd.com).

hjelmn commented:
Best way to proceed would be to run mpi tests with --mca pml ob1 --btl self,sm,uct .. It should work by default on IB but we want to prevent it from using pml/ucx to ensure we test everything. The output of ompi_info -a | grep btl_uct would also be useful to make sure it populates the correct variables.

I also would like to see some tests with rma-mt (github.com/hpc/rma-mt) to see how it performs. I will send you an email with more details.

arun-chandran-edarath commented:
@hjelmn, sorry for the delay in checking the PR in our setup, was super busy in other things.

What I did:

1) git fetch origin pull/13335/head:pr-13335; git checkout pr-13335
2) Built UCX (latest master) and installed to $HOME/openmpi_work/install/oct3/ucx
3) Built Open MPI from this PR:
   a) ./configure --prefix=$HOME/openmpi_work/install/oct3/ompi --with-hwloc=internal --with-libevent=internal --with-pmix=internal CC=gcc CXX=g++ FC=gfortran --without-hcoll --with-xpmem=/home/software/mpi/xpmem/2.7.4 --with-ucx=$HOME/openmpi_work/install/oct3/ucx
   
To check whether I setup things properly, I ran osu latency and bandwidth tests in the below 2 nodes

```
+-----------+                400G InfiniBand (NDR)                +-----------+
| turin003  |=====================================================| turin004  |
| HCA: CX-7 |                                                     | HCA: CX-7 |
+-----------+                                                     +-----------+
```

Observed that sm btl is failing to get loaded.
I have attached all the results and the scripts used. Please check whether the environment and mca parameters are correct.

[result.9526.out.txt](https://github.com/user-attachments/files/22680064/result.9526.out.txt)
[osu_uct_p2p.sbatch.txt](https://github.com/user-attachments/files/22680000/osu_uct_p2p.sbatch.txt)
[result.9526.err.txt](https://github.com/user-attachments/files/22680001/result.9526.err.txt)

bwbarrett commented:
@hjelmn what would you like to do with this PR?  Do we need to bug @bosilca and other Nvidia folks to review?

hjelmn commented:
Rebased on main and fixed some bugs I discovered during additional testing. Also made a change to the  short send method used which may improve small message performance for ob1.
